File: sdk/ruby/examples/oauth-demo/oauth_demo.rb
# GGID OAuth 2.0 Demo (Ruby)
# Run: GGID_URL=... CLIENT_ID=... ruby oauth_demo.rb
require 'webrick'
require 'net/http'
require 'json'
require 'uri'

ggid_url = ENV['GGID_URL'] || 'http://localhost:8080'
client_id = ENV['CLIENT_ID'] || ''
client_secret = ENV['CLIENT_SECRET'] || ''
redirect_uri = ENV['REDIRECT_URI'] || 'http://localhost:3000/callback'

server = WEBrick::HTTPServer.new(Port: 3000)

server.mount_proc '/' do |req, res|
  auth_url = "#{ggid_url}/api/v1/oauth/authorize?" + URI.encode_www_form(
    response_type: 'code', client_id: client_id, redirect_uri: redirect_uri,
    scope: 'openid profile email', state: 'demo'
  )
  res.body = "<h1>GGID OAuth Demo</h1><a href='#{auth_url}'>Login with GGID</a>"
end

server.mount_proc '/callback' do |req, res|
  code = req.query['code']
  uri = URI("#{ggid_url}/api/v1/oauth/token")
  tokens = JSON.parse(Net::HTTP.post_form(uri, {
    grant_type: 'authorization_code', code: code,
    redirect_uri: redirect_uri, client_id: client_id, client_secret: client_secret,
  }).body)

  user_uri = URI("#{ggid_url}/api/v1/oauth/userinfo")
  user_req = Net::HTTP::Get.new(user_uri)
  user_req['Authorization'] = "Bearer #{tokens['access_token']}"
  user = JSON.parse(Net::HTTP.start(user_uri.host, user_uri.port, use_ssl: user_uri.scheme == 'https') { |h| h.request(user_req) }.body)

  res['Content-Type'] = 'application/json'
  res.body = JSON.pretty_generate(tokens: tokens, user: user)
end

trap('INT') { server.shutdown }
server.start

File: sdk/dart/examples/oauth-demo/oauth_demo.dart
// GGID OAuth 2.0 Demo (Dart)
// Run: GGID_URL=... dart run oauth_demo.dart
import 'dart:io';
import 'dart:convert';

void main() async {
  final ggidUrl = Platform.environment['GGID_URL'] ?? 'http://localhost:8080';
  final clientId = Platform.environment['CLIENT_ID'] ?? '';
  final clientSecret = Platform.environment['CLIENT_SECRET'] ?? '';
  final redirectUri = Platform.environment['REDIRECT_URI'] ?? 'http://localhost:3000/callback';

  final server = await HttpServer.bind(InternetAddress.loopbackIPv4, 3000);
  print('OAuth demo on http://localhost:3000');

  await for (final request in server) {
    final path = request.uri.path;
    final response = request.response;

    if (path == '/') {
      final authUrl = '$ggidUrl/api/v1/oauth/authorize?response_type=code'
          '&client_id=$clientId&redirect_uri=${Uri.encodeComponent(redirectUri)}'
          '&scope=openid+profile+email&state=demo';
      response.headers.contentType = ContentType.html;
      response.write('<h1>GGID OAuth Demo</h1><a href="$authUrl">Login with GGID</a>');
    } else if (path == '/callback') {
      final code = request.uri.queryParameters['code'] ?? '';
      final httpClient = HttpClient();
      final tokenReq = await httpClient.postUrl(Uri.parse('$ggidUrl/api/v1/oauth/token'));
      tokenReq.headers.contentType = ContentType.parse('application/x-www-form-urlencoded');
      tokenReq.write('grant_type=authorization_code&code=$code'
          '&redirect_uri=${Uri.encodeComponent(redirectUri)}'
          '&client_id=$clientId&client_secret=$clientSecret');
      final tokenRes = await tokenReq.close();
      final tokens = await tokenRes.transform(utf8.decoder).join();
      response.headers.contentType = ContentType.json;
      response.write(tokens);
      httpClient.close();
    }
    await response.close();
  }
}
